freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserService.java

```diff
@@ -112,39 +123,60 @@ private void asyncSynchronizeUsers(
         Set<String> userCrnFilter, Set<String> machineUserCrnFilter) {
         try {
             Set<String> environmentCrns = stacks.stream().map(Stack::getEnvironmentCrn).collect(Collectors.toSet());
+
+            Optional<String> requestIdOptional = Optional.of(
+                    Optional.ofNullable(MDCBuilder.getMdcContextMap().get(LoggerContextKey.REQUEST_ID.toString()))
+                            .orElseGet(() -> {
+                                String requestId = UUID.randomUUID().toString();
+                                LOGGER.debug("No requestId found. Setting request id to new UUID [{}]", requestId);
+                                MDCBuilder.addRequestId(requestId);
+                                return requestId;
+                            }));
+
+            Json umsEventGenerationIdsJson = userCrnFilter.isEmpty() && machineUserCrnFilter.isEmpty() ?
+                    new Json(umsEventGenerationIdsUtil.getEventGenerationIds(actorCrn, accountId, requestIdOptional)) :
+                    null;
+
             Map<String, UsersState> envToUmsStateMap = umsUsersStateProvider
-                .getEnvToUmsUsersStateMap(accountId, actorCrn, environmentCrns, userCrnFilter, machineUserCrnFilter);
+                .getEnvToUmsUsersStateMap(accountId, actorCrn, environmentCrns, userCrnFilter, machineUserCrnFilter, requestIdOptional);
+
+            Collection<SuccessDetails> success = new ConcurrentLinkedQueue<>();
+            Collection<FailureDetails> failure = new ConcurrentLinkedQueue<>();
 
             Map<String, Future<SyncStatusDetail>> statusFutures = stacks.stream()
                     .collect(Collectors.toMap(Stack::getEnvironmentCrn,
                             stack -> asyncTaskExecutor.submit(() -> {
-                                        MDCBuilder.buildMdcContext(stack);
-                                        return synchronizeStack(stack, envToUmsStateMap.get(stack.getEnvironmentCrn()), userCrnFilter, machineUserCrnFilter);
-                                    })));
-
-            List<SuccessDetails> success = new ArrayList<>();
-            List<FailureDetails> failure = new ArrayList<>();
+                                MDCBuilder.buildMdcContext(stack);
```

Comment:
maybe we should save the MdcContext prior this and restore when finished, otherwise it would contain the last stack's properties which doesn't seem right at first glance

Reply:
The AsyncTaskExecutor is decorated with the MDCCleanerTaskDecorator. That decorator should clean the MDC after each task completes.